### DISCUSSION TITLE
slice() generates infinite bounds

### DISCUSSION THREAD

**Discussion by willisk0612** [pyvista/pyvista#6385] [Q&A]
I am writing a PyVista code for slicing a black/white 3D model. When i use the slice() command, it generates PolyData with infite bounds (1e+299, -1e+299), even though the model it is slicing has finite bounds. Usually only happens on the first few slices of the model, the rest are slices with finite bounds.
Zip file with the used 3D model: [python.zip](https://github.com/user-attachments/files/16250294/python.zip)
```python
import pyvista as pv
import numpy as np

# Load the mesh with materials
mesh = pv.read_meshio('hourglass.obj')

def slice_to_bmp(mesh, color, n_slices):
        if color == 'black':
            color_value = 0
        elif color == 'white':
            color_value = 1

        # Identify each material
        material = mesh.threshold([color_value - 0.1, color_value + 0.1], scalars=mesh.cell_data.active_scalars_name)
    
        #Create a local PyVista plotter object
        plotter = pv.Plotter(off_screen=True)
        plotter.set_background('black')
        
        # Model bounds
        z_min = material.bounds[4]
        z_max = material.bounds[5]

        # Generate a linspace of z-values to slice the mesh. First and last values are the min and max z-values of the mesh
        z_values = np.linspace(z_min, z_max, n_slices)
        print(f'Max material bounds: {max(material.bounds)}')
        for i,z_value in enumerate(z_values):
            slice = material.slice(normal='z', origin=(0, 0, z_value))
            if abs(max(slice.bounds)) > 10000:
                print(f'Slice {i} invalid bounds: {slice.bounds}')
                continue
            
slice_to_bmp(mesh, 'white', 100)
```

### Description
It returns Slice {i} invalid bounds: (1e+299, -1e+299, 1e+299, -1e+299, 1e+299, -1e+299) but is expected to return nothing. 

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Tue Jul 16 16:03:20 2024 Vest-Europa (sommertid)

                OS : Windows (10 10.0.22631 SP0 Multiprocessor FreFie)
            CPU(s) : 8
           Machine : AMD64
      Architecture : 64bit
       Environment : Python
        GPU Vendor : ATI Technologies Inc.
      GPU Renderer : AMD Radeon(TM) Graphics
       GPU Version : 4.6.0 Compatibility Profile Context 24.3.1.240216
  MathText Support : False

  Python 3.11.1 (tags/v3.11.1:a7a450f, Dec  6 2022, 19:58:39) [MSC v.1934 64
  bit (AMD64)]

           pyvista : 0.43.10
               vtk : 9.3.0
             numpy : 1.24.1
        matplotlib : 3.8.3
            scooby : 0.10.0
             pooch : 1.8.2
            pillow : 9.5.0
           imageio : 2.34.2
         pyvistaqt : 0.11.1
             PyQt5 : 5.15.9
           IPython : 8.26.0
        ipywidgets : 8.1.3
             scipy : 1.11.4
              tqdm : 4.66.1
            meshio : 5.3.5
      nest_asyncio : 1.6.0
--------------------------------------------------------------------------------
```

**Comment by user27182** [comment]
Empty meshes are sometimes challenging to handle and work with. In this situation VTK initializes the bounds with these values. Sometimes it's  `(1e+299, -1e+299)`, other times it may just be ones  `(1, -1)` but in either case you can tell they're initialized and not valid bounds based on the signs alone: bounds are specified as `min, max`, but here we can see that `min > max` which cannot be the case for any real bounds Real bounds have `min <= max`. Empty `PolyData`, for example, is initialized with `(1, -1)` bounds:
``` python
import pyvista as pv
pv.PolyData()

#  PolyData (0x1019d0580)
#   N Cells:    0
#   N Points:   0
#   N Strips:   0
#   X Bounds:   1.000e+00, -1.000e+00
#   Y Bounds:   1.000e+00, -1.000e+00
#   Z Bounds:   1.000e+00, -1.000e+00
#   N Arrays:   0
```


> but is expected to return nothing

Would you prefer `None` to be returned in this case? I'm not sure if this is a good solution, since a basic assumption is that the filter returns a dataset, allowing filters to be chained together without having to do if/else checks at every output to make sure we have a mesh to work with at the output.

Maybe a warning for empty outputs would be better? E.g. the `pyvista.Plotter` handle this case
``` python
pv.PolyData().plot()
```
```
ValueError: Empty meshes cannot be plotted. Input mesh has zero points. To allow plotting empty meshes, set `pv.global_theme.allow_empty_mesh = True`
```

Maybe a similar themed config option would help, but for filters in `pyvista.core` instead of for just the `Plotter`? Not sure how or where this would be implemented though, there isn't really a central place to handle this.

Or maybe at the very least we could add a `DataSet.is_empty` property, which could at least provide an easy way to check if the mesh has some geometry to use.

**Reply by willisk0612** [reply]
`but is expected to return nothing` i meant it was expected to print nothing, indicating a sucessfull slice. The problem is that i am losing information about the lower parts of the model when there are invalid slices. I added a print statement at the end of the for-loop, and got this: 
`Slice 0 invalid bounds: (1e+299, -1e+299, 1e+299, -1e+299, 1e+299, -1e+299)
Slice 1 invalid bounds: (1e+299, -1e+299, 1e+299, -1e+299, 1e+299, -1e+299)
Slice 2 invalid bounds: (1e+299, -1e+299, 1e+299, -1e+299, 1e+299, -1e+299)
Slice 3 invalid bounds: (1e+299, -1e+299, 1e+299, -1e+299, 1e+299, -1e+299)
Slice 4 invalid bounds: (1e+299, -1e+299, 1e+299, -1e+299, 1e+299, -1e+299)
Slice 5 invalid bounds: (1e+299, -1e+299, 1e+299, -1e+299, 1e+299, -1e+299)
Slice 6 invalid bounds: (1e+299, -1e+299, 1e+299, -1e+299, 1e+299, -1e+299)
Slice 7 invalid bounds: (1e+299, -1e+299, 1e+299, -1e+299, 1e+299, -1e+299)
Slice 8 sucessful. Z value of slice: 3.313131313131313
Slice 9 sucessful. Z value of slice: 3.7272727272727275
Slice 10 sucessful. Z value of slice: 4.141414141414142
Slice 11 sucessful. Z value of slice: 4.555555555555555`

The lower z-bound of the mesh is z=0, so i have lost the information from z=0 to z=3.31. I am particular interested in getting the slice at z=0, because it may contain important properties that is not present anywhere else. I do not know why VTK initializes the bounds with non-real values when the mesh is not empty at the specified z value? Could it be that it has problems parsing the model?

**Reply by user27182** [reply]
I see, so the issue you're having is not really about how empty meshes are initialized (i.e. that the bounds are `(1e+299, -1e+299)`, but that you believe the mesh should not be empty at all? I.e. that there may be a bug with `slice` returning empty meshes when the mesh should have "something"? Is this correct?

I am not able to run your example locally so I can't really investigate, but it's possible that your dataset isn't defined as you may expect it to be after using `threshold`, e.g. maybe some sections only contain points. `slice` requires cells to slice-through.

E.g. if we modify the example from the [docs for `slice`](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.slice.html#pyvista.DataSetFilters.slice) such that the input dataset (sphere) only has points, the slicing fails.

``` python
import pyvista as pv
sphere = pv.PolyData(pv.Sphere().points)
slice_x = sphere.slice(normal='x')
slice_y = sphere.slice(normal='y')
slice_z = sphere.slice(normal='z')
slices = slice_x + slice_y + slice_z
slices.plot(line_width=5)
```
output:
```
ValueError: Empty meshes cannot be plotted. Input mesh has zero points. To allow plotting empty meshes, set `pv.global_theme.allow_empty_mesh = True`
```

I'd suggest maybe plotting the output after `threshold` before using `slice` to make sure it looks good and has well defined cells in the areas you expect.
